### PR TITLE
Tiler diff layers

### PR DESCRIPTION
**Pull request by rajatSinha303**
# General Information:

**Layer 5** - This layer was initially implemented to visualize data on both sides of the road (left & right) with the help of polygons. The color and thickness (width) of these polygons can be used to display certain information about a road.

**Layer 5 Difference layer** - Layer 5 also has an implementation to visualize the difference in data between two different scenarios (reference & active).

**Layer 51 & 52** - These layers use layer 5's implementation to display data for the reference & active scenarios, but new classes have been implemented for the difference layers.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -44,4 +44,4 @@
 
 	url = https://cornelissenja@bitbucket.org/imbclients/logger-delphi.git
 
-	branch = master
+	branch = master
```
Comment (korenkovv) [inline .gitmodules]: Please revert this file because the only change is the removal of the '\n' symbol.

File: Publish/PublishServerUS.pas [hunk 2]
```diff
@@ -1227,7 +1227,7 @@ function TMetaLayerEntry.CreateUSLayer(aScenario: TScenario; const aTablePrefix:
         diffRange := defaultValue(diffRange, autoDiffRange*0.3);
       end;
     4,    // road color (VALUE_EXPR)
-    5:    // road color (VALUE_EXPR) and width (TEXTURE_EXPR)
+    5, 51, 52:    // road color (VALUE_EXPR) and width (TEXTURE_EXPR)
```
Comment (korenkovv) [inline Publish/PublishServerUS.pas:1230]: "Magic numbers" is no go :) For now this can be left as is but when reimplementing Publisher and Tailer such approach should be banned!

Comment (korenkovv) [inline Publish/PublishServerUS.pas:1230]: For future implementation:
- Each layer type should be represented by a separate entity 
- Each layer has to have it's own class responsible for querying data from the database
- Switch and if-condition on types signals that there are problems in design.

File: TilerServer/TilerWebService.dproj [hunk 3]
```diff
@@ -5,7 +5,7 @@
         <FrameworkType>VCL</FrameworkType>
         <MainSource>TilerWebService.dpr</MainSource>
         <Base>True</Base>
-        <Config Condition="'$(Config)'==''">Release</Config>
+        <Config Condition="'$(Config)'==''">Debug</Config>
```
Comment (korenkovv) [inline TilerServer/TilerWebService.dproj]: Do we want this change?

File: doc/US/Layers/Layer5.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# Tiler Layer 5
```
Comment (korenkovv) [inline doc/US/Layers/Layer5.md]: The documentation is awesome!
The only few thing that I don't agree with:
- Naming. **Layer 5, 51, 52** doesn't tell anything about the layers.
- Inheritance. Doesn't actually make any sense to mention parent-child relation, it doesn't add more info. Naming and grouping should make no need in such kind of relationships between layers.

File: TilerServer/TilerWebModule.pas [hunk 5]
```diff
@@ -568,6 +568,32 @@   TSliceDiffGeometryICLR = class(TSliceDiffOutLineFill)
   protected
     // tile generation
     function GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus; override;
+    function CalculateWidth(aActiveValue, aRefValue: Double; var aValidFlag: Boolean): Double; Virtual;
+    function GetPaletteColor(aActiveTexture, aRefTexture: Double; var aValidFlag: Boolean): TGeoColors;
+    function ConstructPolygon(aPolyPoints: TArray<Double>): TPolygon; Virtual;
+    procedure DrawFillPolygon(aColor: TGeoColors; aBitmap: FMX.Graphics.TBitmap; aPolygon: TPolygon);
+  end;
+  
+  TSliceDiffGeometryICLR2 = class(TSliceDiffGeometryICLR)
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: Name with a number at the end will make the code understandable in the future. Is it possible to give a meaningful name to the class?

File: TilerServer/TilerWebModule.pas [hunk 6]
```diff
@@ -568,6 +568,32 @@   TSliceDiffGeometryICLR = class(TSliceDiffOutLineFill)
...
+  protected
+    // tile generation
+    function GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus; override;
+    function CalculateWidth(aActiveValue, aRefValue: Double; var aValidFlag: Boolean): Double; override;
+    function GetPaletteColor(aActiveTexture, aRefTexture, aWidth: Double; var aValidFlag: Boolean): TGeoColors;
+    function ComputeCoordinateDist(aWidth, aActiveValue, aRefValue, aCapacityFactor, aXY_Diff, aPerpDist: Double; aIsCommonPoly: Boolean): Double;
+    procedure DrawFillPolygon(aColor: TGeoColors; aBitmap: FMX.Graphics.TBitmap; aPolygonCommon, aPolygonExtra: TPolygon; xCommon, xExtra: Double);
+  end;
+
+  TSliceDiffGeometryICLR3 = class(TSliceDiffGeometryICLR)
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: Name with a number at the end will make the code understandable in the future. Is it possible to give a meaningful name to the class?

File: TilerServer/TilerWebModule.pas [hunk 7]
```diff
@@ -3698,69 +3703,518 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
+  widthL := Double.NaN;
+  widthR := Double.NaN;
+  Result := gtsFailed; // sentinel
+  if Assigned(fPalette) then
+  begin
+    aBitmap.Canvas.BeginScene;
+    try
+      aBitmap.Canvas.Clear(0);
+      aBitmap.Canvas.Fill.Kind := TBrushKind.Solid;
+      capacityFactor := 0.001/Abs(aExtent.YMax-aExtent.YMin);
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: Please add a check that ensures YMax and YMin are different.

File: TilerServer/TilerWebModule.pas [hunk 8]
```diff
@@ -3646,39 +3674,16 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
-                then widthL := (isgop.Value.value + refObj.value) / 2
-                else widthL := isgop.Value.value;
-              end
-              else
-              begin
-                if not IsNaN(refObj.value)
-                then widthL := refObj.value
-                else validL := False;
-              end;
+              widthL := CalculateWidth(isgop.Value.value, refObj.value, validL);
```
Comment (AnganaTNO) [inline TilerServer/TilerWebModule.pas]: 'isgop' is does not seem like a good variable name for  CurrentScenario.

File: TilerServer/TilerWebModule.pas [hunk 9]
```diff
@@ -3698,69 +3703,518 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
+begin
+  inherited Create(aLayer, aPalette, aTimeStamp, aCurrentSlice, aRefSlice);
+end;
+
+destructor TSliceDiffGeometryICLR2.Destroy;
+begin
+  inherited;
+end;
+
+function TSliceDiffGeometryICLR2.GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus;
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: This method is too big to easily understand it. Please split it on a number of simpler methods. You already put some comments within the method, this is a very good indicator that you have logical pieces that should be moved into separate methods.

File: TilerServer/TilerWebModule.pas [hunk 10]
```diff
@@ -3698,69 +3703,518 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
+end;
+
+{ TSliceDiffGeometryICLR3 }
+
+constructor TSliceDiffGeometryICLR3.Create(aLayer: TLayer; aPalette: TWDPalette; aTimeStamp: TDateTime; aCurrentSlice, aRefSlice: TSliceGeometryICLR);
+begin
+  inherited Create(aLayer, aPalette, aTimeStamp, aCurrentSlice, aRefSlice);
+end;
+
+function TSliceDiffGeometryICLR3.GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus;
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: Please split this method on smaler parts.

**Review by korenkovv — CHANGES_REQUESTED** [review]

File: TilerServer/TilerWebModule.pas [hunk 11]
```diff
@@ -3646,39 +3674,16 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
-              if not (IsNaN(isgop.Value.texture) or IsNaN(refObj.texture))
-              then colorsL := fPalette.ValueToColors(isgop.Value.texture-refObj.texture)
-              else validL := False;
+              colorsL := GetPaletteColor(isgop.Value.texture, refObj.texture, validL);
+
               //see if we can find R color
-              if not (IsNaN(isgop.Value.texture2) or IsNaN(refObj.texture2))
-              then colorsR := fPalette.ValueToColors(isgop.Value.texture2-refObj.texture2)
-              else validR := False;
+              colorsR := GetPaletteColor(isgop.Value.texture2, refObj.texture2, validR);
```
Comment (AnganaTNO) [inline TilerServer/TilerWebModule.pas]: 'isgop' is not a very good name for ActiveScenario!

**Review by AnganaTNO — CHANGES_REQUESTED** [review]
'isgop' is not a very self-explainatoy name for Active Scenario

File: TilerServer/TilerWebModule.pas (hunk 8 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Fixed!

File: TilerServer/TilerWebModule.pas (hunk 11 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Fixed!

File: TilerServer/TilerWebModule.pas (hunk 5 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Fixed!
TSliceDiffGeometryICLR2 is now TSliceDiffGeometryCommonAndExtraPolygonsLR

File: TilerServer/TilerWebModule.pas (hunk 6 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Fixed!
TSliceDiffGeometryICLR3 is now TSliceDiffGeometryClassChangePolygonStripeLR

File: doc/US/Layers/Layer5.md (hunk 4 above)
Comment (rajatSinha303) [reply doc/US/Layers/Layer5.md]: Thank you :)
Made improvements in the upcoming commit.
Kindly review and suggest changes.

File: .gitmodules (hunk 1 above)
Comment (rajatSinha303) [reply .gitmodules]: Done

File: .gitmodules [hunk 12]
```diff
@@ -45,3 +45,4 @@
 	url = https://cornelissenja@bitbucket.org/imbclients/logger-delphi.git
 
 	branch = master
+
```
Comment (korenkovv) [inline .gitmodules:48]: Please revert this change

File: TilerServer/TilerWebService.dproj (hunk 3 above)
Comment (cornelissenja) [reply TilerServer/TilerWebService.dproj]: We do NOT want this change!

File: TilerServer/TilerWebModule.pas [hunk 13]
```diff
@@ -3627,140 +3657,547 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
+end;
+
+function TSliceDiffGeometryPolygonLR.GetPaletteColor(aActiveTexture, aRefTexture: Double; var aValidFlag: Boolean): TGeoColors;
+begin
+  if not (IsNaN(aActiveTexture) or IsNaN(aRefTexture))then
+    Result := fPalette.ValueToColors(aActiveTexture-aRefTexture)
+  else aValidFlag := False;
+end;
+
+function TSliceDiffGeometryPolygonLR.ConstructPolygon(aPolyPoints: TArray<Double>): TPolygon;
```
Comment (cornelissenja) [inline TilerServer/TilerWebModule.pas]: I think this is very time consuming way to construct a polygon and performance will suffer greatly. Maybe this should be optimized by re-using the same polygon by passing it by ref ("var" in delphi) and because it is always the same amount of points just code them directly?

File: TilerServer/TilerWebService.dproj (hunk 3 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebService.dproj]: Changed back to Release in the upcoming commit

File: .gitmodules (hunk 12 above)
Comment (rajatSinha303) [reply .gitmodules:48]: Done

File: TilerServer/TilerWebModule.pas (hunk 13 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: You are right. Creating another polygon and returning it was not very optimized.
I have passed the polygon by reference and re-used it in the function.
This function (now a procedure, as it does not return anything anymore) was created to remove the redundant code that was initially present for creating polygons in all the three classes for the difference layers.

File: TilerServer/TilerWebModule.pas (hunk 10 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Done!

File: TilerServer/TilerWebModule.pas (hunk 9 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Done!

File: TilerServer/TilerWebModule.pas (hunk 7 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Done!

File: Publish/PublishServerUS.pas (hunk 2 above)
Comment (rajatSinha303) [reply Publish/PublishServerUS.pas:1230]: Noted! 👍

Comment (rajatSinha303) [reply Publish/PublishServerUS.pas:1230]: 👍

File: TilerServer/TilerWebModule.pas [hunk 14]
```diff
@@ -4136,25 +4121,24 @@ function TSliceDiffGeometryPolygonStripeLR.GenerateTileCalc(const aExtent: TExte
...
                 validR := False
-              else
-                widthR := 60;
+              else widthR := 60;
 
               colorsL := GetPaletteColor(curScenarioGeometryObject.texture, refScenarioGeometryObject.texture, widthL, validL);
               colorsR := GetPaletteColor(curScenarioGeometryObject.texture2, refScenarioGeometryObject.texture2, widthR, validR);
             end;
 
-            if not (validR or validL) then //if neither side is valid -> draw a thin black line
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: It might be a good idea if instead of removing comments from the code you could move such code to a separate function with appropriate name.

File: TilerServer/TilerWebModule.pas [hunk 15]
```diff
@@ -584,6 +584,7 @@   TSliceDiffGeometryDoublePolygonLR = class(TSliceDiffGeometryPolygonLR)
     function GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus; override;
     function CalculateWidth(aActiveValue, aRefValue: Double; var aValidFlag: Boolean): Double; override;
     function GetPaletteColor(aActiveTexture, aRefTexture, aWidth: Double; var aValidFlag: Boolean): TGeoColors;
+    procedure PolygonComputation(aXDCommon, aYDCommon, aXDExtra, aYDExtra, aXPrev, aYPrev, aXCurr, aYCurr: Double; aColor: TGeoColors; aBitmap: FMX.Graphics.TBitmap);
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas:587]: Instead of this function you can create two separate functions like "DrawLeftPolygon/DrowRightPolygon" where you can calculate the majority of the arguments you now pass to the "PolygonComputation" and call "DrawFillPolygon" from these functions.

File: TilerServer/TilerWebModule.pas [hunk 16]
```diff
@@ -3894,21 +3887,19 @@ function TSliceDiffGeometryDoublePolygonLR.GenerateTileCalc(const aExtent: TExte
...
+            else
             begin
-              setLength(polygonCommon, 5);
               for part in curScenarioGeometryObject.fGeometry.parts do
               begin
                 xCurr := NaN;
                 yCurr := NaN;
                 for point in part.points do
                 begin
-                  // recalc coordinates relative to extent
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: Can it be a separate function?

File: TilerServer/TilerWebModule.pas [hunk 17]
```diff
@@ -3918,10 +3909,8 @@ function TSliceDiffGeometryDoublePolygonLR.GenerateTileCalc(const aExtent: TExte
                   begin
                     xn := yCurr-yPrev;
                     yn := xPrev-xCurr;
-                    // normalize..
```
Comment (korenkovv) [inline TilerServer/TilerWebModule.pas]: This definitely sounds like a separate function.

**Review by korenkovv — CHANGES_REQUESTED** [review]

File: TilerServer/TilerWebModule.pas (hunk 14 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: I did that. I created a separate function - DrawDefaultPath() which took care of the code inside the if case.
If you mean to say that I should move the whole if condition to a separate function, that would also require creating the 'path' inside that function and hence, passing a lot of parameters for that.

File: TilerServer/TilerWebModule.pas (hunk 16 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: I am sorry but I don't understand what part should be inside a separate function.
Do you mean the following 4 lines of code?
                  xPrev := xCurr;
                  yPrev := yCurr;
                  xCurr := (point.X-aExtent.XMin)/aPixelWidth;
                  yCurr := (aExtent.YMax-point.Y)/aPixelHeight;

File: TilerServer/TilerWebModule.pas (hunk 17 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Yep, you are right. I'll move the distance computation to a separate function.

File: TilerServer/TilerWebModule.pas (hunk 15 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas:587]: I am not sure I understand your point.

In the current procedure prototype for PolygonComputation(), the first 4 parameters are getting calculated just before the procedure is called.
The calculations for each of these 4 variables is done using a common function - ComputeCoordinateDist(aWidth, aActiveValue, aRefValue, aCapacityFactor, aXY_Diff, aPerpDist: Double; aIsCommonPoly: Boolean).

As can be seen, this function already receives 7 arguments.

If we create new functions like "DrawLeftPolygon/DrowRightPolygon" where the first 4 parameters that were passed to PolygonComputation() are computed, a total of about 13 arguments will have to be passed to each of these new functions.

I believe you meant something else compared to what I understood from your comment.
Kindly provide a bit more detailed explanation.

File: TilerServer/TilerWebModule.pas (hunk 17 above)
Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: Done!

File: TilerServer/TilerWebModule.pas [hunk 18]
```diff
@@ -3626,149 +3658,296 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
...
+  if not (IsNaN(aActiveTexture) or IsNaN(aRefTexture))then
+    Result := fPalette.ValueToColors(aActiveTexture-aRefTexture)
+  else aValidFlag := False;
+end;
+
+procedure TSliceDiffGeometryPolygonLR.ConstructPolygon(aPolyPoints: TArray<Double>; var polygon: TPolygon);
+var
+  ele: Integer;
+begin
+  setLength(polygon, 5);
```
Comment (cornelissenja) [inline TilerServer/TilerWebModule.pas]: Not needed anymore?
It is a var and you already set the length in the caller..

Comment (rajatSinha303) [reply TilerServer/TilerWebModule.pas]: You are right. Fixed in the current commit.

**Review by AnganaTNO — APPROVED** [review]
👍

**Review by cornelissenja — APPROVED** [review]
A approve of this message, uhhhh of this merge request :-)
